### ISSUE TITLE
Support for run highlights

### ISSUE THREAD

**Issue by MWorkman99** [mwilliamson/python-mammoth#122]
The w:highlight docx element could easily be supported in the Boolean case. Thus, is there a highlight on the text or not.

This is inline with how underline currently works (and small_caps, i, b, etc..), and thus probably not outside the scope of Mammoth.

Example docx code:

`<w:r ><w:rPr><w:szCs w:val="12"/><w:highlight w:val="yellow"/></w:rPr><w:t>3031.3-SE2898</w:t></w:r>`

custom_styles could be used to map the highlight to html elements and styles.

I have working code for the Boolean case, if this is approved.

**Comment by abdnh** [comment]
I'm interested in seeing support for highlighting. Could you publish the code (maybe in a PR) please?

**Comment by mwilliamson** [comment]
Could you post a minimal example document, the expected HTML, and the actual HTML?

**Comment by abdnh** [comment]
OK. I got a working version with support for the boolean case (which is sufficient for my purposes):
```diff
diff --git a/mammoth/conversion.py b/mammoth/conversion.py
index bdc8687..43aeb81 100644
--- a/mammoth/conversion.py
+++ b/mammoth/conversion.py
@@ -125,6 +125,8 @@ class _DocumentConverter(documents.element_visitor(args=1)):
             paths.append(self._find_style_for_run_property("italic", default="em"))
         if run.is_bold:
             paths.append(self._find_style_for_run_property("bold", default="strong"))
+        if run.is_highlight:
+            paths.append(html_paths.element(["mark"], fresh=False))
         paths.append(self._find_html_path_for_run(run))

         for path in paths:
diff --git a/mammoth/documents.py b/mammoth/documents.py
index 7559d40..0ba7b69 100644
--- a/mammoth/documents.py
+++ b/mammoth/documents.py
@@ -50,6 +50,7 @@ class Run(HasChildren):
     is_strikethrough = cobble.field()
     is_all_caps = cobble.field()
     is_small_caps = cobble.field()
+    is_highlight = cobble.field()
     vertical_alignment = cobble.field()
     font = cobble.field()
     font_size = cobble.field()
@@ -125,6 +126,7 @@ def run(
     is_strikethrough=None,
     is_all_caps=None,
     is_small_caps=None,
+    is_highlight=None,
     vertical_alignment=None,
     font=None,
     font_size=None,
@@ -141,6 +143,7 @@ def run(
         is_strikethrough=bool(is_strikethrough),
         is_all_caps=bool(is_all_caps),
         is_small_caps=bool(is_small_caps),
+        is_highlight=bool(is_highlight),
         vertical_alignment=vertical_alignment,
         font=font,
         font_size=font_size,
diff --git a/mammoth/docx/body_xml.py b/mammoth/docx/body_xml.py
index c11e9e5..9e56d50 100644
--- a/mammoth/docx/body_xml.py
+++ b/mammoth/docx/body_xml.py
@@ -94,6 +94,7 @@ def _create_reader(numbering, content_types, relationships, styles, docx_file, f
         is_strikethrough = read_boolean_element(properties.find_child("w:strike"))
         is_all_caps = read_boolean_element(properties.find_child("w:caps"))
         is_small_caps = read_boolean_element(properties.find_child("w:smallCaps"))
+        is_highlight = read_boolean_element(properties.find_child("w:highlight"))

         def add_complex_field_hyperlink(children):
             hyperlink_kwargs = current_hyperlink_kwargs()
@@ -115,6 +116,7 @@ def _create_reader(numbering, content_types, relationships, styles, docx_file, f
                 is_strikethrough=is_strikethrough,
                 is_all_caps=is_all_caps,
                 is_small_caps=is_small_caps,
+                is_highlight=is_highlight,
                 vertical_alignment=vertical_alignment,
                 font=font,
                 font_size=font_size,
```
(tests not updated)

Here is a sample document to test with:
[highlight-sample.docx](https://github.com/mwilliamson/python-mammoth/files/9599702/highlight-sample.docx)


Result before the change:
```HTML
<p>This is highlighted text.</p>
```

Result after the change:
```HTML
<p>This is <mark>highlighted</mark> text.</p>
```

**Comment by MWorkman99** [comment]
Here is my diff file to accomplish this goal.


```diff
diff -ruN mammoth/conversion.py ../../mrepo/mammoth/conversion.py
--- mammoth/conversion.py	2022-09-19 19:55:07.004041200 -0400
+++ ../../mrepo/mammoth/conversion.py	2022-06-01 15:59:42.715286100 -0400
@@ -123,6 +123,8 @@
             paths.append(html_paths.element(["sup"], fresh=False))
         if run.is_italic:
             paths.append(self._find_style_for_run_property("italic", default="em"))
+        if run.is_highlight:
+            paths.append(self._find_style_for_run_property("highlight", default="yellow"))
         if run.is_bold:
             paths.append(self._find_style_for_run_property("bold", default="strong"))
         paths.append(self._find_html_path_for_run(run))
@@ -355,7 +357,7 @@
 
 
 def _document_matcher_matches(matcher, element, element_type):
-    if matcher.element_type in ["underline", "strikethrough", "all_caps", "small_caps", "bold", "italic", "comment_reference"]:
+    if matcher.element_type in ["underline", "strikethrough", "all_caps", "small_caps", "bold", "italic", "highlight", "comment_reference"]:
         return matcher.element_type == element_type
     elif matcher.element_type == "break":
         return (
diff -ruN mammoth/document_matchers.py ../../mrepo/mammoth/document_matchers.py
--- mammoth/document_matchers.py	2022-09-19 19:55:07.007031600 -0400
+++ ../../mrepo/mammoth/document_matchers.py	2022-06-01 15:59:42.718279700 -0400
@@ -50,6 +50,9 @@
 class small_caps(object):
     element_type = "small_caps"
 
+class highlight(object):
+    element_type = "highlight"
+
 
 class comment_reference(object):
     element_type = "comment_reference"
diff -ruN mammoth/documents.py ../../mrepo/mammoth/documents.py
--- mammoth/documents.py	2022-09-19 19:55:07.011022600 -0400
+++ ../../mrepo/mammoth/documents.py	2022-06-01 15:59:42.721271300 -0400
@@ -50,6 +50,7 @@
     is_strikethrough = cobble.field()
     is_all_caps = cobble.field()
     is_small_caps = cobble.field()
+    is_highlight = cobble.field()
     vertical_alignment = cobble.field()
     font = cobble.field()
     font_size = cobble.field()
@@ -125,6 +126,7 @@
     is_strikethrough=None,
     is_all_caps=None,
     is_small_caps=None,
+    is_highlight=None,
     vertical_alignment=None,
     font=None,
     font_size=None,
@@ -141,6 +143,7 @@
         is_strikethrough=bool(is_strikethrough),
         is_all_caps=bool(is_all_caps),
         is_small_caps=bool(is_small_caps),
+        is_highlight=bool(is_highlight),
         vertical_alignment=vertical_alignment,
         font=font,
         font_size=font_size,

diff -ruN mammoth/docx/body_xml.py ../../mrepo/mammoth/docx/body_xml.py
--- mammoth/docx/body_xml.py	2022-09-19 19:55:07.018002500 -0400
+++ ../../mrepo/mammoth/docx/body_xml.py	2022-06-01 15:59:42.728253600 -0400
@@ -94,6 +94,7 @@
         is_strikethrough = read_boolean_element(properties.find_child("w:strike"))
         is_all_caps = read_boolean_element(properties.find_child("w:caps"))
         is_small_caps = read_boolean_element(properties.find_child("w:smallCaps"))
+        is_highlight = read_boolean_element(properties.find_child("w:highlight"))
 
         def add_complex_field_hyperlink(children):
             hyperlink_kwargs = current_hyperlink_kwargs()
@@ -115,6 +116,7 @@
                 is_strikethrough=is_strikethrough,
                 is_all_caps=is_all_caps,
                 is_small_caps=is_small_caps,
+                is_highlight=is_highlight,
                 vertical_alignment=vertical_alignment,
                 font=font,
                 font_size=font_size,

diff -ruN mammoth/styles/parser/document_matcher_parser.py ../../mrepo/mammoth/styles/parser/document_matcher_parser.py
--- mammoth/styles/parser/document_matcher_parser.py	2022-09-19 19:55:07.095796000 -0400
+++ ../../mrepo/mammoth/styles/parser/document_matcher_parser.py	2022-06-01 15:59:42.824992900 -0400
@@ -52,6 +52,9 @@
     elif tokens.try_skip(TokenType.IDENTIFIER, "small-caps"):
         return document_matchers.small_caps
 
+    elif tokens.try_skip(TokenType.IDENTIFIER, "highlight"):
+        return document_matchers.highlight
+        
     elif tokens.try_skip(TokenType.IDENTIFIER, "comment-reference"):
         return document_matchers.comment_reference
 

```

**Comment by MalikRumi** [comment]
uhhhh....hey folks, what happened here? Does the patch work? Pass tests? Added to repo? Rejected? Inquiring minds want to know.

**Comment by mwilliamson** [comment]
Support for style mappings for highlights has been released as part of 1.8.0.
